src/runtime/contrib/tidl/tidl_runtime.cc

```diff
@@ -0,0 +1,248 @@
+
+// #define TVM_RUNTIME_DBG_TIDL_TIMER
+#ifdef TVM_RUNTIME_DBG_TIDL_TIMER
+struct timespec t0, t1;
+#define tick() clock_gettime(CLOCK_MONOTONIC, &t0);
+#define tock() \
+  (clock_gettime(CLOCK_MONOTONIC, &t1), t1.tv_sec - t0.tv_sec + (t1.tv_nsec - t0.tv_nsec) / 1e9)
+#endif
+
+extern "C" {
+extern void TidlRunSubgraph(
+    int total_subgraphs,     // total number of subgraphs to be executed on TIDL
+    int subgraph_id,         // subgraph id: 0, 1, 2, etc.
+    int batch_size,          // batch value of input tensor
+    int num_inputs,          // number of inputs to the subgraph
+    int num_outputs,         // number of outputs of the subgraph
+    float** inputTensors,    // array of input tensors
+    float** outputTensors);  // array of output tensors
+}
+typedef void (*tidl_subgraph_t)(int, int, int, int, int, float**, float**);
+
+inline std::string ToJSON(int total_subgraphs,
+                          const std::unordered_map<std::string, int>& num_inputs_,
+                          const std::unordered_map<std::string, int>& num_outputs_) {
+  std::ostringstream os;
+  dmlc::JSONWriter writer(&os);
+  writer.BeginObject();
+  writer.WriteObjectKeyValue("total subgraphs", total_subgraphs);
+  writer.WriteObjectKeyValue("subgraph inputs", num_inputs_);
+  writer.WriteObjectKeyValue("subgraph outputs", num_outputs_);
+  writer.EndObject();
+  return os.str();
+}
+
+inline void FromJSON(const std::string& str, int* total_subgraphs,
+                     std::unordered_map<std::string, int>* num_inputs,
+                     std::unordered_map<std::string, int>* num_outputs) {
+  std::istringstream is(str);
+  dmlc::JSONReader reader(&is);
+  dmlc::JSONObjectReadHelper helper;
+  // Read total subgraphs
+  helper.DeclareField("total subgraphs", total_subgraphs);
+  // Read num_inputs
+  helper.DeclareField("subgraph inputs", num_inputs);
+  // Read num_outputs
+  helper.DeclareField("subgraph outputs", num_outputs);
+  helper.ReadAllFields(&reader);
+}
+
+namespace tvm {
+namespace runtime {
+
+/*! \brief A module for TIDL runtime. */
+class TIDLModule : public runtime::ModuleNode {
+ public:
+  explicit TIDLModule(int total_subgraphs, const std::unordered_map<std::string, int>& num_inputs,
+                      const std::unordered_map<std::string, int>& num_outputs) {
+    this->total_subgraphs_ = total_subgraphs;
+    this->num_inputs_ = num_inputs;
+    this->num_outputs_ = num_outputs;
+    this->tidl_handle = NULL;
+  }
+
+  /*!
+   * \brief Initialize TIDL runtime by loading subgraph execution function from
+   * TIDL library.
+   */
+  void TidlInit() {
+    if (!tidl_handle) {
+      // Load TIDL shared library
+      dlerror();
+      tidl_handle = dlopen("libtidl_api.so", RTLD_NOW | RTLD_GLOBAL);
+      const char* dlsym_error1 = dlerror();
+      if (dlsym_error1) {
+        LOG(FATAL) << "Cannot open libtidl_api.so! " << dlsym_error1 << '\n';
+      }
+      // Load TIDL subgraph execution function
+      dlerror();
```

Comment:
Is this necessary, same as above.

Reply:
https://linux.die.net/man/3/dlerror: the correct way to test for an error is to call dlerror() to clear any old error conditions, then call dlsym(), and then call dlerror() again, saving its return value into a variable, and check whether this saved value is not NULL